File: 2017/cpp/day05.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
  vector<int> jumps, jumps2;
  int n;
  while (cin >> n) {
    jumps.push_back(n);
    jumps2.push_back(n);
  }

  int steps = 0;
  int idx = 0;
  while (idx >= 0 && idx < int(jumps.size())) {
    int jump = jumps[idx];
    jumps[idx] += 1;
    idx += jump;
    steps++;
  }

  cout << "Part 1: " << steps << '\n';

  steps = 0;
  idx = 0;
  while (idx >= 0 && idx < int(jumps2.size())) {
    int jump = jumps2[idx];
    if (jump >= 3) {
      jumps2[idx] -= 1;
    } else {
      jumps2[idx] += 1;
    }
    idx += jump;
    steps++;
  }

  cout << "Part 2: " << steps << '\n';
}

File: 2017/rust/src/bin/day05.rs
use std::io;
use std::io::prelude::*;

fn main() {
    let mut jumps: Vec<i32> = vec![];
    for line in io::stdin().lock().lines() {
        jumps.push(line.unwrap().parse().unwrap());
    }

    let mut steps = 0;
    let mut idx: i32 = 0;
    while idx >= 0 && idx < jumps.len() as i32 {
        let instruction = &mut jumps[idx as usize];
        let offset = *instruction;
        if offset >= 3 {
            *instruction -= 1;
        } else {
            *instruction += 1;
        }
        idx += offset;
        steps += 1;
    }

    println!("{}", steps);
}
